src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ThemeManager3.cs

```diff
@@ -0,0 +1,434 @@
+using System.Diagnostics;
+
+
+namespace System.Windows;
+
+internal static class ThemeManager3
+{
+
+
+    #region Internal Methods
+
+    internal static void OnSystemThemeChanged()
+    {
+        if(IsFluentThemeEnabled)
+        {
+            IgnoreAppResourcesChange = true;
+
+            bool useLightColors = GetUseLightColors(Application.Current.ThemeMode);
+            var fluentThemeResourceUri = GetFluentThemeResourceUri(useLightColors);
+            AddOrUpdateThemeResources(Application.Current.Resources, fluentThemeResourceUri);
+
+            foreach(Window window in Application.Current.Windows)
+            {
+                if(window.ThemeMode == ThemeMode.None)
+                {
+                    ApplyStyleOnWindow(window, useLightColors);
+                }
+                else
+                {
+                    ApplyFluentOnWindow(window);
+                }
+            }
+
+            IgnoreAppResourcesChange = false;
+        }
+        else
+        {
+            foreach(Window window in FluentEnabledWindows)
+            {
+                if(window == null || window.IsDisposed) continue;
+
+                if(window.ThemeMode == ThemeMode.None)
+                {
+                    RemoveFluentFromWindow(window);
+                }
+                else
+                {
+                    ApplyFluentOnWindow(window);
+                }
+            }
+        }
+    }
+
+    internal static void OnApplicationThemeChanged(ThemeMode oldThemeMode, ThemeMode newThemeMode)
+    {
+        IgnoreAppResourcesChange = true;
+
+        try
+        {
+            if(newThemeMode == ThemeMode.None)
+            {
+                if(oldThemeMode != newThemeMode)
+                {
+                    RemoveFluentFromApplication();
+                }
+                return;
+            }
+
+            bool useLightColors = GetUseLightColors(newThemeMode);
+            var fluentThemeResourceUri = GetFluentThemeResourceUri(useLightColors);
+            AddOrUpdateThemeResources(Application.Current.Resources, fluentThemeResourceUri);
+
+            foreach(Window window in Application.Current.Windows)
+            {
+                // Replace this with a check for the window theme
+                if(!FluentEnabledWindows.HasItem(window))
+                {
+                    ApplyStyleOnWindow(window, useLightColors);
+                }
+            }
+        }
+        finally
+        {
+            IgnoreAppResourcesChange = false;
+        }
+    }
+
+    internal static void OnWindowThemeChanged(Window window, ThemeMode oldThemeMode, ThemeMode newThemeMode)
+    {
+        if(newThemeMode == ThemeMode.None)
+        {
+            if(newThemeMode != oldThemeMode)
+            {
+                RemoveFluentFromWindow(window);
+            }
+            return;
+        }
+
+        ApplyFluentOnWindow(window);
+    }
+
+    internal static bool SyncThemeModeAndResources()
```

Comment:
Should add `if(Application.Current == null) return;` here?

Reply:
This is only called from Application.InvalidateResourceReference, therefore I didn't add the null check